src/plonk/permutation.rs

```diff
@@ -94,3 +100,76 @@ impl Assembly {
         Ok(())
     }
 }
+
+#[derive(Clone, PartialEq, Default)]
+pub(crate) struct PermutationData {
+    columns: Box<[Column<Any>]>,
+    mapping: Box<[Vec<(usize, usize)>]>,
+}
+
+impl PermutationData {
+    pub fn new<F: PrimeField>(cs: &ConstraintSystem<F>, perm_assembly: &Assembly) -> Self {
+        Self {
+            columns: cs.permutation().get_columns().into_boxed_slice(),
+            mapping: perm_assembly.mapping.clone().into_boxed_slice(),
+        }
+    }
+
+    pub fn matrix<F: PrimeField>(
+        &self,
+        k_table_size: usize,
+        num_io: &[usize],
+        num_advice_columns: usize,
+    ) -> SparseMatrix<F> {
+        util::construct_permutation_matrix(
+            k_table_size,
+            num_io,
+            &self.columns,
+            num_advice_columns,
+            &self.mapping,
+        )
+    }
+}
+
+impl Serialize for PermutationData {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
```

Comment:
where will the serialization be used?

Reply:
We use it in the public parameter hash calculation. It used to have a permutation matrix in there, but now we have this strcuture in there

Note: We can't just 'derive' serialization, since the `Column` from halo2 is not serialized